src/args.rs

```diff
@@ -0,0 +1,441 @@
+use core_affinity::CoreId;
+use either::Either;
+use serde::{
+	de::{Deserialize, SeqAccess, Visitor},
+	*,
+};
+use thiserror::Error;
+
+use crate::params::{CpuCount, GuestMemorySize};
+
+/*
+ * Few notes:
+ * Although clap offers the convenient default_value_t macro, which would otherwise
+ * help us avoid using Option<T> everywhere, this was removed from structs like
+ * UhyveArgs, CpuArgs, etc., for two reasons:
+ *
+ * 1. toml-rs does not have a similar macro, instead choosing to rely on Option<T>.
+ * 2. The structs, derived from both CLI arguments (clap) and TOML configuration files
+ *    (toml-rs), are fed to Params. Params implements a Default type itself.
+ *
+ * There is no need to have the structs representing the command line or configuration
+ * file interface decide on defaults, as such defaults are decided upon by Params.
+ * Params is ultimately what defines, or should define, how Uhyve will actually work.
+ * Params should contain a composition of configurations collected both by clap and
+ * by toml-rs.
+ *
+ * Both interfaces should only strictly require a configuration if determining a
+ * default is not possible, e.g. the location of a kernel. In the future, it might be
+ * possible to infer a certain default from one interface but not for the other (e.g.
+ * using the pathname of a TOML configuration file to substitute the `kernel` argument.)
+ */
+
+/// This is the config that defines a set of parameters for a given Hermit machine image.
+///
+/// Note that the field names are relevant for the TOML's tables.
+///
+/// TODO: Figure out how to adapt file isolation.
+#[derive(Debug, Deserialize)]
+pub struct UhyveGuestConfig {
+	pub memory: MemoryArgs,
+	pub cpu: CpuArgs,
+	pub guest: GuestArgs,
+}
+
+use merge::Merge;
+
+#[derive(Default, Merge, Parser, Debug, Deserialize)]
+pub struct MemoryArgs {
+	/// Guest RAM size
+	#[clap(short = 'm', long, env = "HERMIT_MEMORY_SIZE")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub memory_size: Option<GuestMemorySize>,
+
+	/// Disable ASLR
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub no_aslr: Option<bool>,
+
+	/// Transparent Hugepages
+	///
+	/// Advise the kernel to enable Transparent Hugepages [THP] on the virtual RAM.
+	///
+	/// [THP]: https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub thp: Option<bool>,
+
+	/// Kernel Samepage Merging
+	///
+	/// Advise the kernel to enable Kernel Samepage Merging [KSM] on the virtual RAM.
+	///
+	/// [KSM]: https://www.kernel.org/doc/html/latest/admin-guide/mm/ksm.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub ksm: Option<bool>,
+}
+
+/// Arguments for the CPU resources allocated to the guest.
+#[derive(Default, Debug, Clone, Merge, Parser, Deserialize)]
+pub struct CpuArgs {
```

Comment:
AFAIK you can specify the merge strategy per struct, btw.:
https://docs.rs/merge_derive/0.2.0/src/merge_derive/lib.rs.html#44

Reply:
setting the struct strategy to `#[serde(default)]` so that entire sections of the TOML configuration can be omitted